openedx_learning/core/publishing/model_mixins.py

```diff
@@ -0,0 +1,264 @@
+        ---------------
+        Note that because we're just using the underlying model's relations,
+        calling this a second time will returned the cached relation, and
+        not cause a fetch of new data from the database. So for instance, if
+        you do::
+
+            # Create a new Component + ComponentVersion
+            component, component_version = create_component_and_version(
+                learning_package_id=learning_package.id,
+                namespace="xblock.v1",
+                type="problem",
+                local_key="monty_hall",
+                title="Monty Hall Problem",
+                created=now,
+                created_by=None,
+            )
+
+            # This will work, because it's never been published
+            assert component.versioning.published is None
+
+            # Publishing happens
+            publish_all_drafts(learning_package.id, published_at=now)
+
+            # This will FAIL because it's going to use the relation value
+            # cached on component instead of going to the database again.
+            # You need to re-fetch the component for this to work.
+            assert component.versioning.published == component_version
+
+            # You need to manually refetch it from the database to see the new
+            # publish status:
+            component = get_component_by_pk(component.pk)
+
+            # Now this will work:
+            assert component.versioning.published == component_version
+
+        TODO: This probably means we should use a custom Manager to select
+        related fields.
+        """
+
+        def __init__(self, content_obj):
+            self.content_obj = content_obj
+
+            self.content_version_model_cls = (
+                PublishableContentModelRegistry.get_versioned_model_cls(
+                    type(content_obj)
+                )
+            )
+            # Get the field that points from the *versioned* content model
+            # (e.g. ComponentVersion) to the PublishableEntityVersion.
+            field_to_pev = self.content_version_model_cls._meta.get_field(
+                "publishable_entity_version"
+            )
+
+            # Now that we know the field that leads to PublishableEntityVersion,
+            # get the reverse related field name so that we can use that later.
+            self.related_name = field_to_pev.related_query_name()
+
+        @property
+        def draft(self):
+            """
+            Return the content version object that is the current draft.
+
+            This will not force re-fetches from the database. Please read the
+            "Caching Warning" in the class docstring for more details.
+            """
+            pub_ent = self.content_obj.publishable_entity
+
+            # If no Draft object exists at all, then no PublishableEntityVersion was
+            # ever created (just the PublishableEntity)
+            if not hasattr(pub_ent, "draft"):
```

Comment:
Sorry, I think I'm confused, how does the `draft` attribute get assigned on the PublishableEntity instance?

Reply:
That's the OneToOneField relation, but I replaced it with a straight query on the `Draft` model, which forces the refresh and is probably a lot clearer.